### DISCUSSION TITLE
Transaction exceeded timeout of 20s

### DISCUSSION THREAD

**Discussion by klipitkas** [planetscale/discussion#610] [Bug Reports]
Hello Planetscale folks, I hope you are all doing well, we are using Laravel (using Eloquent as the ORM of choice) + PHP PDO and are seeing some unusual things when it comes to planetscale. For example we have the following error:
```
SQLSTATE[HY000]: General error: 1105 target: production.-.primary: vttablet: rpc error: code = Aborted desc = transaction 1707253104872624490: ended at 2024-02-13 21:05:49.154 UTC (exceeded timeout: 20s) (CallerID: vejcja4pih4qbc4w76ty) (SQL: select * from `listing_property_definitions` where `listing_property_definitions`.`deleted_at` is null)
```

Checking sentry we see the following:
<img width="535" alt="image" src="https://github.com/planetscale/discussion/assets/3259834/13b69a52-69d4-4bf0-b94b-7d0671ddff91">

This error seems to be caused by a simple select query, in a table that has 50 entries:
<img width="1128" alt="image" src="https://github.com/planetscale/discussion/assets/3259834/2ab7064c-475e-47be-b790-63cae56d6115">

The error does not make much sense, we also seeing in that timeframe some `UPDATE` queries that fail silently while no exceptions are thrown from the ORM because it returns that the update worked correctly however the changes are not persisted in the DB.

**Comment by orware** [comment]
Hello @klipitkas,

While this wouldn't be common for PHP, by any chance might you have `autocommit` disabled for your connections in any way?

Typically this issue isn't that common for our PHP users to encounter and is typically only found when users are in fact running long-running transactions exceeding out 20 second transaction timeout limit, but it's not something you should see with a simple `SELECT` query.

If your connections are for sure using the default of having `autocommit` enabled, then you would probably want to look further up the callstack to what calls the `prepareForValidation()` method and see if there is a larger batch operation being run within a transaction since that is going to be the likely cause of the 20 second transaction timeout limit being reached. If that is the case, then making some adjustments there on the application side to reduce the batch size running within the transaction will probably be the likely solution.

Depending on how the above suggestions work out, if you'd prefer to work with us more directly feel free to write in over at https://support.planetscale.com/ and we can more easily gather additional details, look at logs on our end, etc. that may provide further clues.

**Reply by klipitkas** [reply]
Hey @orware thanks for the quick answer, I did some checks and it seems that PDO has autocommit enabled:
```
> $pdo = DB::connection()->getPdo();
= PDO {#7991
    inTransaction: false,
    attributes: {
      CASE: NATURAL,
      ERRMODE: EXCEPTION,
      AUTOCOMMIT: 1,
      PERSISTENT: false,
      DRIVER_NAME: "mysql",
      SERVER_INFO: b"ÿ\x17\x04#08S01command handling not implemented yet: 9",
      ORACLE_NULLS: NATURAL,
      CLIENT_VERSION: "mysqlnd 8.1.27",
      SERVER_VERSION: "8.0.34-PlanetScale",
      STATEMENT_CLASS: [
        "PDOStatement",
      ],
      EMULATE_PREPARES: 0,
      CONNECTION_STATUS: "aws.connect.psdb.cloud via TCP/IP",
      DEFAULT_FETCH_MODE: BOTH,
    },
  }
```

I will have a look and if I don't find anything with batches in a transaction I will try to contact support. Thanks again.
